mlir/lib/Transforms/RemoveDeadValues.cpp

```diff
@@ -169,7 +169,10 @@ static SmallVector<OpOperand *> operandsToOpOperands(OperandRange operands) {
 /// Here, cleaning means:
 ///   (1) Dropping all its uses, AND
 ///   (2) Erasing it
-/// iff it has no memory effects and none of its results are live.
+/// iff:
+///   (1) It has no memory effects
+///   (2) None of its results are live
+///   (3) It is not used as an argument in a public function call
```

Comment:
Why does it matter whether the function is private or public?

Reply:
You can't touch a function API when you don't see all the calls, and by definition a public function has external calls you don't see.

You may clone the function to specialize it though, but that's a different tradeoff to make